src/v/cluster/self_test/netcheck.cc

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/self_test/netcheck.h"
+
+#include "cluster/logger.h"
+#include "utils/gate_guard.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+
+#include <boost/range/irange.hpp>
+
+namespace cluster::self_test {
+
+void netcheck::validate_options(const netcheck_opts& opts) {
+    if (opts.peers.empty()) {
+        throw netcheck_exception("Internal setting, 'peers' must not be empty");
+    }
+    if (opts.request_size <= 0 || opts.request_size > (1ULL << 27)) {
+        throw netcheck_option_out_of_range(
+          "Request size must be > 0 and < 128MB");
+    }
+    if (opts.parallelism <= 0 || opts.parallelism > 256) {
+        throw netcheck_option_out_of_range(
+          "Parallelism option must be > 0 and <= 256");
+    }
+    const auto duration = std::chrono::duration_cast<std::chrono::seconds>(
+      opts.duration);
+    if (duration < 1s || opts.duration > (5 * 60s)) {
+        throw netcheck_option_out_of_range(
+          "Duration must be between 1 second and 5 minutes");
+    }
+}
+
+netcheck::netcheck(
+  model::node_id self, ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _connections(connections) {}
+
+ss::future<> netcheck::start() { return ss::now(); }
+
+ss::future<> netcheck::stop() {
+    _as.request_abort();
+    return _gate.close();
+}
+
+void netcheck::cancel() { _cancelled = true; }
+
+ss::future<std::vector<self_test_result>> netcheck::run(netcheck_opts opts) {
+    vassert(_gate.get_count() == 0, "Benchmark already in progress");
+    gate_guard g{_gate};
+    _cancelled = false;
+    _opts = opts;
+    try {
+        co_await ss::futurize_invoke(validate_options, opts);
+        vlog(
+          clusterlog.info,
+          "Starting redpanda self-test network benchmark, with options: {}",
+          opts);
+        co_return co_await ss::with_scheduling_group(opts.sg, [this]() {
+            return ssx::async_transform(
+              _opts.peers, [this](model::node_id peer) {
+                  return run_individual_benchmark(peer);
+              });
+        });
+    } catch (const netcheck_exception& ex) {
+        vlog(clusterlog.error, "Network exception encountered: {}", ex);
+        throw;
+    } catch (const std::exception& ex) {
+        vlog(clusterlog.error, "Unforseen exception encountered: {}", ex);
+        throw netcheck_exception(ex.what());
```

Comment:
In general, when we see an unexpected exception type, it's better to let it bubble up to a higher level rather than changing the type and rethrowing -- e.g. in this case do_start_test already knows how to handle a std::exception and include its .what() in the results, so this branch isn't needed.